custom/icds_reports/utils/aggregation.py

```diff
@@ -0,0 +1,194 @@
+from __future__ import absolute_import
+import hashlib
+
+from dateutil.relativedelta import relativedelta
+from toggle.models import Toggle
+
+from corehq.apps.userreports.models import StaticDataSourceConfiguration, get_datasource_config
+from corehq.apps.userreports.util import get_table_name
+
+
+class BaseICDSAggregationHelper(object):
+    """Defines an interface for aggregating data from UCRs to specific tables
+    for the dashboard.
+
+    All aggregate tables are partitioned by state and month
+
+    Attributes:
+        ucr_data_source_id - The UCR data source that contains the raw data to aggregate
+        aggregate_parent_table - The parent table defined in models.py that will contain aggregate data
+        aggregate_child_table_prefix - The prefix for tables that inherit from the parent table
+    """
+    ucr_data_source_id = None
+    aggregate_parent_table = None
+    aggregate_child_table_prefix = None
+
+    def __init__(self, state_id, month):
+        self.state_id = state_id
+        self.month = month
+
+    @property
+    def domain(self):
+        # Currently its only possible for one domain to have access to the ICDS dashboard per env
+        return Toggle.get('dashboard_icds_reports').enabled_users[0][len('domain:'):]
+
+    @property
+    def ucr_tablename(self):
+        doc_id = StaticDataSourceConfiguration.get_doc_id(self.domain, self.ucr_data_source_id)
+        config, _ = get_datasource_config(doc_id, self.domain)
+        return get_table_name(self.domain, config.table_id)
+
+    def generate_child_tablename(self, month=None):
+        month = month or self.month
+        month_string = month.strftime('%Y-%m')
+        hash_for_table = hashlib.md5(self.state_id + month_string).hexdigest()[8:]
+        return self.aggregate_child_table_prefix + hash_for_table
+
+    def create_table_query(self, month=None):
+        month = month or self.month
+        month_string = month.strftime('%Y-%m-%d')
+        tablename = self.generate_child_tablename(month)
+
+        return """
+        CREATE TABLE IF NOT EXISTS \"{child_tablename}\" (
+            CHECK (month = '{month_string}' AND state_id = '{state_id}'),
+            LIKE \"{parent_tablename}\" INCLUDING DEFAULTS INCLUDING CONSTRAINTS INCLUDING INDEXES
+        ) INHERITS (\"{parent_tablename}\")
+        """.format(
```

Comment:
String formatting this way is vulnerable to SQL injection, although when you control the source (know types/content format) of all input values it's possibly not as much of a concern. On the other hand, evolution of code over time (sources of input change, for example) can allow vulnerabilities to sneak in undetected. It's probably necessary for at least some of the values such as the table names to be inserted using string formatting. However, the typical/safer way to insert values like `month_string` and `state_id` is to use bind parameters. We should prefer bind parameters whenever possible.

This comment also applies to other queries constructed below as well.

Reply:
:+1: makes sense to me. I originally decided not to do that since there will be no user input for these, but agree this is safer. https://github.com/dimagi/commcare-hq/pull/19521/commits/d33ee7896626d2453369dd589edc3acd9e3e3505

Also happy to offline if you have any better ways to organize this. I tried to keep all database access contained in the models and all query generation in the helper. Could also do sql scripts, but it's been a pain in the past keeping up with multiple SQL scripts whereas this will hopefully be nicer in that all the logic is in one place for each table